# Linear Optimization

Optimization techniques are widely used in various business applications to improve efficiency, reduce costs, maximize profits, and make better decisions. Some common business applications of optimization techniques include:

1. <span style='color:blue'>Supply Chain Management:</span> Optimizing inventory levels, production schedules, and distribution networks to minimize costs and improve customer service.

2. <span style='color:blue'>Resource Allocation:</span> Optimizing the allocation of resources such as manpower, machinery, and materials to maximize productivity and minimize waste.

3. <span style='color:blue'>Transportation and Logistics:</span> Optimizing vehicle routes and schedules to minimize fuel consumption, reduce transportation costs, and improve delivery times.

4. <span style='color:blue'>Financial Planning:</span> Optimizing investment portfolios, risk management strategies, and financial models to maximize returns and minimize risks.

5. <span style='color:blue'>Marketing and Pricing:</span> Optimizing marketing campaigns, pricing strategies, and product placements to maximize revenue and market share.

6. <span style='color:blue'>Production Planning:</span> Optimizing production processes, scheduling, and capacity utilization to meet demand while minimizing costs.

7. <span style='color:blue'>Scheduling and Timetabling:</span> Optimizing employee schedules, project timelines, and event schedules to improve efficiency and reduce conflicts.

From this list we will use No. 2. Optimizing the allocation of resources for the machiner.  The following scenario depicts.  

A manufacturing company produces three products: Product A, Product B, and Product C. Each product requires different amounts of two types of machines, Machine X and Machine Y, as shown in the table below:
Product	Machine X hours/unit	Machine Y hours/unit
A	                   2	              4
B	                   1	              3
C	                   3	              2

Machine X is available for 100 hours per week, and Machine Y is available for 85 hours per week. The company wants to maximize its total production output.
a.	Please formulate this problem as a linear programming problem in standard form. 
b.	Solve the problem in Excel.
c.	The problem is bounded by machine X or Y? If you have a budget to upgrade the machine, will you upgrade machine X or Y? Please explain your answer by
    exploring the math model. 

### Linear Programming Standard Form

Let:

x be the number of units of Product A produced per week.
𝑦 be the number of units of Product B produced per week.
𝑧 be the number of units of Product C produced per week.

Given:

Machine X is available for 100 hours per week.
Machine Y is available for 85 hours per week.

Objective: Maximize total production output (maximize 𝑥+𝑦+𝑧).

Constraints:

1.  For Machine X: 2𝑥+𝑦+3𝑧≤100
    Each unit of Product A requires 2 hours of Machine X.
    Each unit of Product B requires 1 hour of Machine X.
    Each unit of Product C requires 3 hours of Machine X.
    Therefore, if the company produces 𝑥 units of Product A, 
    𝑦 units of Product B, and z units of Product C, then the total hours required on Machine X is 2𝑥+𝑦+3𝑧

Since Machine X is only available for 100 hours per week, the total hours needed cannot exceed 100. This gives us the constraint.

2.  For Machine Y: 4𝑥+3𝑦+2𝑧≤85
    Each unit of Product A requires 4 hours of Machine Y.
    Each unit of Product B requires 3 hours of Machine Y.
    Each unit of Product C requires 2 hours of Machine Y.
    Thus, the total hours required on Machine Y for producing 
    𝑥 units of Product A, y units of Product B, and z units of Product C is 4𝑥+3𝑦+2𝑧
    4x+3y+2z.

Machine Y is available for only 85 hours per week, so the hours used cannot exceed 85. This gives us the constraint.

3. Non-negativity constraints: 𝑥,𝑦,𝑧≥0
These constraints ensure that the production quantities 𝑥, 𝑦, and 𝑧 are non-negative since it’s impossible to produce a negative quantity of a product.

This yields the following linear programming formulation:

Maximize: 𝑥+𝑦+𝑧

subject to:

2𝑥+𝑦+3𝑧≤100
4𝑥+3𝑦+2𝑧≤85
𝑥,𝑦,𝑧≥0








In [27]:
# Import the PuLP library

from pulp import LpProblem, LpMaximize, LpVariable, lpSum, LpStatus, value

# Create the linear programming problem
model = LpProblem("Maximize_Production_Output", LpMaximize)

# Decision variables
x_A = LpVariable('x_A', lowBound=0, cat='Continuous')  # Units of Product A
x_B = LpVariable('x_B', lowBound=0, cat='Continuous')  # Units of Product B
x_C = LpVariable('x_C', lowBound=0, cat='Continuous')  # Units of Product C

# Objective function
model += lpSum([x_A, x_B, x_C]), "Total_Production_Output"

# Constraints
model += 2 * x_A + 1 * x_B + 3 * x_C <= 100, "Machine_X_Constraint"
model += 4 * x_A + 3 * x_B + 2 * x_C <= 85, "Machine_Y_Constraint"

# Solve the problem
model.solve()

# Output the results
print("Status:", LpStatus[model.status])
print("Produce units of Product A:", value(x_A))
print("Produce units of Product B:", value(x_B))
print("Produce units of Product C:", value(x_C))
print("Total Production Output:", value(model.objective))

Status: Optimal
Produce units of Product A: 0.0
Produce units of Product B: 7.8571429
Produce units of Product C: 30.714286
Total Production Output: 38.5714289


#### Linear Programming Model Breakdown

1.  Status: Optimal.  The model found a solution that will maximizes the objective function and meet all constraints.
2.  Produce units of Product A: 0.0: is not beneficial to produce Product A under current constraints, this is likely due to the machine time limitation for
    Machine X (2 hours) and Y (4 hours)
4.  Produce units of Product B: 7.86: Optimal to produce about 7.86 units of Product B, working within machine constraints.
5.  Produce units of Product C: 30.71: Optimal to produce about 30.71 units of Prodct C, utilizing machine hours more efficiently
6.  Total Production Output: 38.57: Combined production of B and C, maximizing output without producing Product A.

### Solve the Problem in Excel.  A Separate file will be sent.

### The Problem is bounded by machine X or Y? 
If you have a budget to upgrade the machine, will you upgrade machine X or Y? Please explain your answer by exploring the math model.

Both machines are binded based off the constraints in this scenario.  The folowing should be considered.

#### Decision for Upgrading

1.  Impact of Upgrading each machine:
   
    Upgrading Machine X: Since it's fully utilized, increasing its available hours could allow for producing more units of Products B and C, as they
    contribute significantly to total output.
   
    Upgrading Machine Y: Similarly, upgrading Machine Y will also allow for an increase in production since it's also at capacity.

2.  Comparative Benefit:

    To determine which machine to upgrade, consider which machine has more potential for increasing production. In this case:
    The production times per unit for each product may suggest that Machine Y's upgrades could yield a greater increase in output, especially given that
    Product B and C both require relatively less time from Machine Y compared to Machine X.
    
    Additionally, if you observe that production is currently constrained by Machine X or if a significant portion of the output would be affected by
    increasing Machine Y's hours, you may prioritize the upgrade of Machine Y.

#### Recommendation

If there is a budget for upgrading, the analysis indicates that either machine upgrade could potentially yield increased output, but you should favor 
upgrading the machine that is:
    
    More heavily utilized in production (both are, but analyze specific needs of production levels).
    Has more significant production time savings for the products with higher outputs.
    
In practice, you could also conduct sensitivity analysis to explore how varying the availability of machine hours impacts the optimal production levels,
which would provide additional insights into which machine to upgrade.

When we explore the math model in this linear model, each machine's constraint is represented by an inequality based of the hours.  Here are the constraints:
Machine X Constraint:

2𝑥𝐴+1𝑥𝐵+3𝑥𝐶≤100

This constraint means that the total time Products A, B, and C require on Machine X cannot exceed 100 hours.

Each unit of Product A takes 2 hours, Product B takes 1 hour, and Product C takes 3 hours on Machine X.

In the optimal solution, we calculated that Machine X uses exactly 100 hours, so it’s a binding constraint (meaning it’s fully utilized and limiting 
production).

Machine Y Constraint:

4𝑥𝐴+3𝑥𝐵+2𝑥𝐶≤85

This constraint means the total time spent on Machine Y by all products cannot exceed 85 hours.

Each unit of Product A requires 4 hours, Product B takes 3 hours, and Product C takes 2 hours on Machine Y.

In the optimal solution, Machine Y also uses exactly 85 hours, making it another binding constraint.

Since both constraints are binding, they both limit production output. Therefore, upgrading either machine will allow more units of Products B and C to be produced, as both machines are fully utilized.

#### Deciding Which Machine to Upgrade

To determine which machine to upgrade, let’s analyze:

As noted above, the Time Requirements Per Unit: Products B and C have relatively balanced usage across both machines, but upgrading Machine Y might yield a more favorable outcome because it has slightly lower time requirements per unit for the higher-output products (B and C).
Final Recommendation
An upgrade to either Machine X or Y would result in increased production, but if forced to choose one based on this model:

Upgrading Machine Y may provide a slightly greater flexibility in production increases due to the lower time requirements for Products B and C, which contribute the most to the optimal production output.

In summary, analyzing the math model indicates that while both machines are at capacity, upgrading Machine Y could offer a marginally better impact on increasing production output given its use across the high-output products in this setup.



## Conclusion:  

The solution tells us to make zero product of A, and about 8 of product B, and 31 product C.  Product A is not feasible with the limited resource it is in.  If we focus on B and will be more productive for the machines.  Unless we decide to upgrade.
